gvfs-helper.c

```diff
@@ -0,0 +1,2424 @@
+
+		strbuf_release(&pattern);
+	}
+
+	strbuf_release(&config_data);
+
+	if (match) {
+		gh__global.cache_server_url = p_url;
+		trace2_data_string("gvfs-helper", NULL, "cache/url", p_url);
+	}
+
+	else if (gh__cmd_opts.cache_server_mode ==
+		 GH__CACHE_SERVER_MODE__VERIFY_ERROR) {
+		if (status.ec != GH__ERROR_CODE__OK)
+			error("config: %s", status.error_message.buf);
+		die("could not verify cache-server '%s'", p_url);
+	}
+
+	else if (gh__cmd_opts.cache_server_mode ==
+		 GH__CACHE_SERVER_MODE__VERIFY_DISABLE) {
+		if (status.ec != GH__ERROR_CODE__OK)
+			warning("config: %s:", status.error_message.buf);
+		warning("could not verify cache-server '%s'", p_url);
+		trace2_data_string("gvfs-helper", NULL, "cache/url",
+				   "disabled");
+	}
+
+	gh__response_status__release(&status);
+}
+
+/*
+ * Read stdin until EOF (or a blank line) and add the desired OIDs
+ * to the oidset.
+ *
+ * We expect stdin to contain output from a command like:
+ * rev-list --objects [--missing=print] [--filter-print-omitted]
+ *
+ * Stdin should contain a list of OIDs.  It may have additional
+ * decoration that we need to strip out.
+ *
+ * We expect:
+ * <hex_oid> [<path>]   // present OIDs
+ * ~<hex_oid>           // omitted OIDs
+ * ?<hex_oid>           // missing OIDs
+ *
+ * Allowing this slightly "dirty" format saves us from having to
+ * do something like "rev-list | grep | sed | gvfs-helper".
+ */
+static unsigned long read_stdin_from_rev_list(struct oidset *oids, int b_missing_only)
+{
+	struct object_id oid;
+	struct strbuf buf_stdin = STRBUF_INIT;
+	unsigned long count = 0;
+
+	do {
+		const char *p = NULL;
+
+		if (strbuf_getline(&buf_stdin, stdin) == EOF || !buf_stdin.len)
+			break;
+
+		if (buf_stdin.buf[0] == '?')
+			p = buf_stdin.buf + 1;
+		else if (b_missing_only)
+			continue;
+		else if (buf_stdin.buf[0] == '~')
+			p = buf_stdin.buf + 1;
+		else
+			p = buf_stdin.buf;
+
+		if (get_oid_hex(p, &oid))
+			continue; /* just silently eat it */
+
+		if (!oidset_insert(oids, &oid))
+			count++;
+	} while (1);
+
+	return count;
+}
+
+/*
+ * Build a complete JSON payload for a gvfs/objects POST request
+ * containing the first n OIDs in an OIDSET index by the iterator.
+ *
+ * https://github.com/microsoft/VFSForGit/blob/master/Protocol.md
+ */
+static unsigned long build_json_payload__gvfs_objects(
+	struct json_writer *jw_req,
+	struct oidset_iter *iter,
+	unsigned long nr_in_block)
+{
+	unsigned long k;
+	const struct object_id *oid;
+
+	k = 0;
+
+	jw_init(jw_req);
+	jw_object_begin(jw_req, 0);
+	jw_object_intmax(jw_req, "commitDepth", gh__cmd_opts.depth);
+	jw_object_inline_begin_array(jw_req, "objectIds");
+	while (k < nr_in_block && (oid = oidset_iter_next(iter))) {
+		jw_array_string(jw_req, oid_to_hex(oid));
+		k++;
+	}
+	jw_end(jw_req);
+	jw_end(jw_req);
+
+	return k;
+}
+
+/*
+ * Lookup the creds for the main/origin Git server.
+ *
+ * Both VFS and Scalar always require authentication for both
```

Comment:
FYI - VFS4G and Scalar [do support anonymous access](https://github.com/microsoft/VFSForGit/blob/339b6a62b43501dee21f3dd491055e4bfeecdd92/GVFS/GVFS.Common/Git/GitAuthentication.cs#L173) with the main Git server.  The test repo used for functional tests (https://gvfs.visualstudio.com/ci/_git/ForTests) is configured to allow anonymous access.

It is still the case that the cache-servers always require authentication.

(I don't know this code well enough yet to know if that matters 😄 )

Reply:
I just pushed a fix for this.  Thanks!